File: 39-combination-sum/combination-sum.cpp
class Solution {
public:
    void generateCombinations(int index,vector<int>& candidates,int target,vector<int>&subsets,vector<vector<int>>&ans)
    {

        if(target==0)
        {
            ans.push_back(subsets);
            return;
        }
        if(index==candidates.size() || target<0)
        {
            return;
        }
        if(target-candidates[index]>=0)
        {
            subsets.push_back(candidates[index]);
            generateCombinations(index,candidates,target-candidates[index],subsets,ans);
            subsets.pop_back();
        }
        generateCombinations(index+1,candidates,target,subsets,ans);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) 
    {
        vector<int>subsets;
        vector<vector<int>>ans;

        generateCombinations(0,candidates,target,subsets,ans);
        return ans;
    }
};

File: 39-combination-sum/combination-sum.java
class Solution 
{
    public void findTotalSubsets(int index,int[] nums,int n,List<List<Integer>>ans,List<Integer>subset,int target)
    {
        if(target==0)
        {
            ans.add(new ArrayList<>(subset));
            return;
        }
        if(index==n || target<0)
        {   
            return;
        }
        subset.add(nums[index]);
        findTotalSubsets(index,nums,n,ans,subset,target-nums[index]);
        subset.remove(subset.size()-1);
        findTotalSubsets(index+1,nums,n,ans,subset,target);
        
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) 
    {
        List<List<Integer>> ans=new ArrayList<>();
        List<Integer> subset= new ArrayList<>();

        findTotalSubsets(0,candidates,candidates.length,ans,subset,target);
        return ans;
    }
}